46239 v2 (no runtime layout changes)

trylek opened the pull request:
This is a simplified version of my original fix for 46239 as the full fix including runtime layout cleanup,

https://github.com/dotnet/runtime/pull/51416

has an enormous bug tail and seems too risky to take relatively late in the release cycle.

Thanks

Tomas

/cc @dotnet/crossgen-contrib

trylek committed:
Fix regression test 46239 and improve runtime logging

The regression test

<code>src\tests\JIT\Regressions\JitBlue\Runtime_46239</code>

exercises various interesting corner cases of type layout that
weren't handled properly in Crossgen2 on x86 and ARM[32]. This
change fixes the remaining deficiencies and it also adds
provisions for better runtime logging upon type layout mismatches.

Thanks

Tomas

trylek committed:
Switch over the runtime outerloop pipeline to use Crossgen2

With this change, the only remaining pipelines using Crossgen1 are
"r2r.yml", "r2r-extra.yml" and "release-tests.yml". I haven't yet
identified the pipeline running the "release-tests.yml" script;
for the "r2r*.yml", these now remain the only pipelines exercising
Crossgen1. I don't think it makes sense to switch them over to
CG2 as we already have their CG2 counterparts; my expectation is
that, once CG1 is finally decommissioned, they will be just deleted.

Thanks

Tomas

trylek committed:
One more fix for explicit layout with zero fields

trylek committed:
Remove no longer needed R2RMFLA override of IsBlittableOrManagedSequential

jkotas commented:
Since we agreed that this is just a stop-gap measure, I think it would better to omit the last commit that introduces a bunch of undesirable dependencies, like making ILVerify depend on interop marshalling.

trylek commented:
@jkotas - Hmm, so am I right to understand that you're suggesting I should basically undo the moves of marshalling-related files originally suggested by Michal and implement the BlittableOrManagedSequential check using a virtual override in ReadyToRunMetadataFieldLayoutAlgorithm?

MichalStrehovsky commented:
> @jkotas - Hmm, so am I right to understand that you're suggesting I should basically undo the moves of marshalling-related files originally suggested by Michal and implement the BlittableOrManagedSequential check using a virtual override in ReadyToRunMetadataFieldLayoutAlgorithm?

I agree with Jan - since having interop intertwined with type layout is not our long term plan per the conversation in the other pull request (and we want to get your other change in after .NET 6 cuts off), it doesn't make sense to move interop into type system just to undo that later this year.

trylek committed:
Implement IsBlittableOrManagedSequential using a R2R-specific override

jkotas commented on src/coreclr/tools/Common/TypeSystem/Common/MetadataFieldLayoutAlgorithm.cs:
```diff
@@ -349,6 +352,7 @@ protected static ComputedInstanceFieldLayout ComputeExplicitFieldLayout(Metadata
                     );
                 if (needsToBeAligned)
                 {
+                    largestAlignmentRequired = LayoutInt.Max(largestAlignmentRequired, type.Context.Target.LayoutPointerSize);
```
I would expect that `ComputeFieldSizeAndAlignment` will return sufficient alignment for GC pointers or anything with GC pointers. Is this really needed? If yes, it would be worth a comment explaining the situation where it is needed.

trylek replied:
I believe you're right, I have replaced this with just an assert double-checking that at this point <code>largestAlignmentRequired</code> must be greater than or equal to <code>PointerSize</code>.

jkotas commented on src/coreclr/tools/Common/TypeSystem/Common/MetadataFieldLayoutAlgorithm.cs:
```diff
@@ -304,17 +305,19 @@ protected virtual void FinalizeRuntimeSpecificStaticFieldLayout(TypeSystemContex
             cumulativeInstanceFieldPos -= offsetBias;
 
             var layoutMetadata = type.GetClassLayout();
+            LayoutInt instanceSize = cumulativeInstanceFieldPos + new LayoutInt(layoutMetadata.Size) + offsetBias;
```
`ComputeInstanceSize` is bumps this up to `layoutMetadata.Size` again.

I assume that this needs to be here because `ComputeInstanceSize` applies it on top of `InstanceByteCountUnaligned`, but this one applies it on top `InstanceByteCount`? A comment may be useful since this does not make sense by just looking at the code.

trylek replied:
Thanks for pointing that out, auditing this aspect has revealed a few rough edges around the base class size calculation; I'm now locally testing an update cleaning this up by consistently using <code>CalculateFieldBaseOffset</code> in all the relevant places, I'll push the extra commit once I make sure it works reasonably well.

jkotas commented on src/coreclr/tools/Common/TypeSystem/Common/MetadataFieldLayoutAlgorithm.cs:
```diff
@@ -304,17 +305,19 @@ protected virtual void FinalizeRuntimeSpecificStaticFieldLayout(TypeSystemContex
         }
 
-        protected static ComputedInstanceFieldLayout ComputeExplicitFieldLayout(MetadataType type, int numInstanceFields)
+        protected virtual bool IsBlittableOrManagedSequential(TypeDesc type) => throw new NotImplementedException();
```
Nit: Maybe change this to `protected virtual bool AlignUpInstanceByteSizeForExplicitFieldLayoutCompatQuirk(TypeDesc type) => true;`, so that we do not have to worry about implementing it in other type system instances.

trylek committed:
Rename the method per JanK's PR feedback

trylek commented:
OK, I believe the initial results aren't the worst, the test 46239 still fails on arm but otherwise the Pri1 runs seem mostly clean. I'm working on retesting the change after cleaning up base class size manipulation per JanK's PR feedback and I'm going to look into the arm issue but I believe the change is mostly functional.

trylek committed:
Fix two explicit layout details

1) Mimic CoreCLR runtime bug filed under #53542;

2) Fix pointer alignment for byref-like fields.

trylek committed:
Fix bad conflict merge

trylek committed:
Mark remaining CG2 outerloop failures to clean up the CI results

trylek committed:
Remove artificial alignment change per DavidWr's PR feedback

trylek committed:
Fix the StructPacking test on ARM

trylek committed:
Improve fix for StructPacking, finally fix 46239 on ARM

trylek committed:
Further refining of the ARM case

trylek committed:
Fix subtle difference in x86 alignment with vs. without layout

trylek commented:
OK, I finally have a green outerloop run (including the CG2 legs). I believe the change is ready to be merged in.

jkotas commented on src/coreclr/tools/Common/TypeSystem/Common/MetadataFieldLayoutAlgorithm.cs:
```diff
@@ -304,18 +305,26 @@ protected virtual void FinalizeRuntimeSpecificStaticFieldLayout(TypeSystemContex
+            LayoutInt cumulativeInstanceFieldPos = CalculateFieldBaseOffset(type, requiresAlign8: false, requiresAlignedBase: false) - offsetBias;
+            if (!type.IsValueType)
+            {
+                // CoreCLR runtime has a bug in field offset calculation in the presence of class inheritance
```
What is a simple repro to demonstrate this bug?

I have tried:
```csharp
using System;
using System.Runtime.InteropServices;

[StructLayout(LayoutKind.Explicit)]
class A1
{
   [FieldOffset(0)]
   public int x;
}

[StructLayout(LayoutKind.Explicit)]
class A2 : A1
{
   [FieldOffset(0)]
   public int y;
}

[StructLayout(LayoutKind.Explicit)]
class A3 : A2
{
   [FieldOffset(0)]
   public int z;
}

class My
{
    static unsafe void Main()
    {
        A3 a = new A3();
        fixed (int *px = &a.x)
        fixed (int *pz = &a.z)
        {
            Console.WriteLine(pz - px);
        }
    }
}
```
It prints `2` for me. I would expect it to print more than that given this bug that doubles the base offsets.

trylek replied:
When I modify your example like this:

<pre>
using System;
using System.Runtime.InteropServices;

[StructLayout(LayoutKind.Explicit)]
class A1
{
   [FieldOffset(0)]
   public byte x;
   [FieldOffset(2)]
   public char x2;
}

[StructLayout(LayoutKind.Explicit)]
class A2 : A1
{
   [FieldOffset(0)]
   public int y;
}

[StructLayout(LayoutKind.Explicit)]
class A3 : A2
{
   [FieldOffset(0)]
   public byte z;
}

class My
{
    static unsafe void Main()
    {
        A3 a = new A3();
        fixed (byte *px = &a.x)
        fixed (byte *pz = &a.z)
        {
            Console.WriteLine(pz - px);
        }
    }
}
</pre>

I receive 24; if properly packed, I think it should say 8 like you said. Your example doesn't exhibit the issue because it hits this weird statement I stumbled upon a couple of times when working on the change:

https://github.com/dotnet/runtime/blob/01a8f497a0fb02aac891ae400fd59b5de65c0244/src/coreclr/vm/methodtablebuilder.cpp#L8563

Thanks

Tomas

trylek replied:
I guess I should probably add this condition to the quirk, will do.

jkotas replied:
> I receive 24; 

And it prints 19 on .NET Framework and .NET Core 3.1. It looks like we accidentally changing behavior here. @jkoritzinsky Can it be caused by your field layout refactoring? It would be nice to get this fixed and covered by tests.

jkotas replied:
> I guess I should probably add this condition to the quirk, will do.

I do not think we need to maintain this quirk. I would delete this, disable the test that it hitting this corner case and file a bug to get it fixed. Note that fixing this is not going to be R2R breaking change.

jkoritzinsky replied:
Possibly. I think I changed a little bit about how classes with inheritance are considered blittable/non-blittable. That might have changed this.

trylek replied:
OK, I'll remove the quirk and I'll retest to see what exactly breaks.

trylek replied:
https://github.com/dotnet/runtime/issues/53542

trylek committed:
Remove the duplicate base type size quirk per JanK's PR feedback

trylek commented:
Removal of the quirk naturally hit one of the typesystem tests. I'll fix that once the outerloop tests finish along with any potential additions to issues.targets for tests affected by this issue.

trylek committed:
Update the typesystem unit tests to removal of the base size quirk

trylek commented:
OK, I have removed the double base size quirk and we're discussing the fix on the issue thread https://github.com/dotnet/runtime/issues/53542. My previous outerloop testing was green so I'm not planning to run a new outerloop test as that doesn't exercise the Crossgen2 unit testing. Please let me know if there's anything else that needs addressing before merging this long-standing change in.

davidwrighton reviewed (approved):
I've been following this change, and it looks good to me now.

jkotas reviewed (approved):
LGTM as well. Thank you!

mangod9 commented:
Thanks @trylek on fixing this!
